R/intervals_support_funs.R

```diff
@@ -0,0 +1,363 @@
+#' All new intervals created will be added right after their original ones.
+#' @return A modified PKNCAdata object with specified imputation methods on the target intervals.
+#' @examples
+#' d_conc <- data.frame(
+#'   conc = c(1, 0.6, 0.2, 0.1, 0.9, 0.4, 1.2, 0.8, 0.3, 0.2, 1.1, 0.5),
+#'   time = rep(0:5, 2),
+#'   ID = rep(1:2, each = 6),
+#'   analyte = rep(c("Analyte1", "Analyte2"), each = 6)
+#' )
+#'
+#' d_dose <- data.frame(
+#'   dose = c(100, 200),
+#'   time = c(0, 0),
+#'   ID = c(1, 2)
+#' )
+#'
+#' library(PKNCA)
+#' o_conc <- PKNCA::PKNCAconc(d_conc, conc ~ time | ID / analyte)
+#' o_dose <- PKNCA::PKNCAdose(d_dose, dose ~ time | ID)
+#'
+#' intervals <- data.frame(
+#'   start = c(0, 0, 0),
+#'   end = c(3, 5, Inf),
+#'   half.life = c(TRUE, TRUE, TRUE),
+#'   cmax = c(TRUE, TRUE, TRUE),
+#'   impute = c("start_conc0,start_predose", "start_predose", "start_conc0"),
+#'   analyte = c("Analyte1", "Analyte2", "Analyte1")
+#' )
+#'
+#' o_data <- PKNCAdata(o_conc, o_dose, intervals = intervals)
+#'
+#' # Apply interval_add_impute function
+#' o_data <- interval_add_impute(o_data,
+#'                               target_impute = "start_conc0",
+#'                               target_params = "half.life",
+#'                               target_groups = data.frame(analyte = "Analyte1"))
+#' @export
+interval_add_impute <- function(data, ...) {
+  UseMethod("interval_add_impute", data)
+}
+
+#' @export
+interval_add_impute.PKNCAdata <- function(data, target_impute, after = Inf,
+                                          target_params = NULL, target_groups = NULL, ...) {
+  # If the impute column is not present, add it to the intervals
+  if (!"impute" %in% names(data$intervals) && !is.null(data$impute)) {
+    data$intervals$impute <- data$impute
+    data$impute <- NA_character_
+  }
+  data$intervals <- interval_add_impute.data.frame(data$intervals, target_impute,
+                                                   after, target_params, target_groups)
+  data
+}
+
+#' Add impute method to the impute column
+#'
+#' This is an internal helper function used to add an impute method to the impute column.
+#'
+#' @param impute_vals Character vector of impute methods.
+#' @param target_impute The imputation method to be added.
+#' @param after Numeric value specifying the index position in which to add the impute.
+#' @return A character string or vector with the added impute method.
+#' @keywords internal
+add_impute_method <- function(impute_vals, target_impute, after) {
+  # Make sure the character vector has length
+  if (length(impute_vals) == 0) return(impute_vals)
+
+  # Remove the impute from the other methods in each value
+  impute_vals <- ifelse(is.na(impute_vals), "", impute_vals)
+  strsplit(impute_vals, split = "[ ,]+") |>
+    lapply(FUN = setdiff, target_impute) |>
+    lapply(FUN = append, values = target_impute, after = after) |>
+    vapply(FUN = paste, collapse = ",", FUN.VALUE = "")
+}
+
+#' @export
```

Comment:
issue: No need to export this function, I would use `@keywords internal` instead.\
```suggestion
#' @keywords internal
```